From 285fa13139772b936e6664720695408d4c1c74c2 Mon Sep 17 00:00:00 2001
From: Arnon Kaufman <akaufman@habana.ai>
Date: Thu, 15 May 2025 12:55:42 +0300
Subject: [PATCH] [SW-206854] g3: gt: numa: fix core scan loop

Missing CPU's core on the NUMA-cores' affiliate list (option topo
--cpu). On GT setups the NUMA zones are spreed interlaced, different
from the development setups.

Change-Id: I23d697081c7cc6aaf1fefb1075f18a8f4626290e
Signed-off-by: Arnon Kaufman <akaufman@habana.ai>
---

diff --git a/app/hl-smi/info.cpp b/app/hl-smi/info.cpp
index 5a9c00d..a6206d8 100644
--- a/app/hl-smi/info.cpp
+++ b/app/hl-smi/info.cpp
@@ -1768,12 +1768,29 @@
 	for (i = 0; i < max; i++) {
 		if (set[i / __BITS_PER_LONG] &
 		    ((unsigned long)1 << ( i % __BITS_PER_LONG))) {
-			for (j = i + 1; j < max; j++) {
+
+			/* SW-206854, need ti fix scenario were the last CPU
+			 * is singular, and not part of a range.
+			 * instead of refactor the code, a tiny change is needed
+			 * to include the last CPU as part of the exist
+			 * calculation logic. Done by allow an extra pass in
+			 * order to call the print_affinity_range. Bare in mind
+			 * that the in case the CPU is part of a range it still
+			 * needed to be supported.
+			 * range scan - by simulate a clear bit at max + 1,
+			 * the range calculation is done correctly for both
+			 * single and range.
+			 */
+			for (j = i + 1; j <= max; j++) {
 				/* if the bit is "set, and it's not the last
-				 * value - continue! */
-				if ((set[j / __BITS_PER_LONG] &
-				    ((unsigned long)1 << ( j % __BITS_PER_LONG))) &&
-				    j + 1 != max)
+				 * value - continue!
+				 * to simulate a clr bit and still avoid invalid
+				 * memory access, check for max + 1, and avoid
+				 * the check. Skipping the continue equivalent
+				 * to having a clr bit
+				 */
+				if ((j < max) && set[j / __BITS_PER_LONG] &
+				    ((unsigned long)1 << ( j % __BITS_PER_LONG)))
 					continue;
 
 				/* reached "cleared" bit OR reached the end -
